include/maliput/math/kd_tree.h

```diff
@@ -154,75 +212,43 @@ class KDTree {
   };
 
   // Functor for comparing points according to the given dimension being evaluated at that point.
-  struct NodeCmp {
-    NodeCmp(std::size_t index) : index_(index) {
-      MALIPUT_VALIDATE(index < Dimension, "Index can not be greater than number of dimensions minus one.");
-    }
-    // Compares two nodes according to the given dimension being evaluated at that point.
-    bool operator()(const Node& n1, const Node& n2) const { return n1.get(index_) < n2.get(index_); }
-
-    std::size_t index_;
-  };
-
-  // Makes a tree from a range of points in the data structure.
-  // @param begin Is the start of range.
-  // @param end Is the end of range.
-  // @param index Is the dimension being evaluated.
-  Node* MakeTree(std::size_t begin, std::size_t end, std::size_t index) {
-    // If range is empty, no tree is needed to be built.
-    if (end <= begin) return nullptr;
-    const std::size_t n = begin + (end - begin) / 2;
-    auto i = nodes_.begin();
-    // Sorting the element in the middle of the range(median).
-    // Smaller and greater values will be located to the left and right of the range correspondingly, according to
-    // NodeCmp functor. However, those values aren't sorted.
-    std::nth_element(i + begin, i + n, i + end, NodeCmp(index));
-    // Obtaining the index to be used for sorting in the next call to MakeTree.
-    index = (index + 1) % Dimension;
-    nodes_[n].set_left(MakeTree(begin, n, index));
-    nodes_[n].set_right(MakeTree(n + 1, end, index));
-    return &nodes_[n];
-  }
 
   // Obtains the nearest point in the @p node to the given @p point.
-  // The @p index under evaluation is provided as this method is called recursively.
-  // Updates the #visited, #best_, and #best_dist_ variables.
-  void Nearest(const typename KDTree<Coordinate, Dimension>::Node* node, const Coordinate& point,
-               std::size_t index) const {
+  // @param node The node to be evaluated.
+  // @param point The point to be evaluated.
+  // @param index Dimension under evaluation as this method is called recursively.
+  // @param nearest_neighbour_node The nearest neighbour node so far.
+  // @param nearest_neighbour_distance The closest distance to the nearest neighbour so far.
+  void Nearest(const Node* node, const Coordinate& point, std::size_t index, double tolerance,
+               Node*& nearest_neighbour_node, double* nearest_neighbour_distance) const {
```

Comment:
Node*&?!'+- -> this will lead to confusion ;)

What do you think about changing the return type to Node*?

Reply:
> What do you think about changing the return type to Node*?

Think recursively: as it goes deeper and deeper I want that pointer to be changing in tandem. I can't have this behavior with a return type.